internal/provider/resource_application.go

```diff
@@ -668,6 +709,59 @@ func (r *applicationResource) Create(ctx context.Context, req resource.CreateReq
 	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
 }
 
+// createCharmResources processes the resources map specified
+// and combines it with information on image registries.
+// Each resource can be either a revision number or an OCI image URL.
+// If the resource is a revision number, it is used as the charmRevision.
+// If the resource is an OCI image URL, it is used as the ociImageURL.
+// If the OCI image URL's registry matches one in the imageRegistries map,
+// the corresponding username and password are used for authentication.
+func createCharmResources(planResources map[string]string, imageRegistries map[string]registryDetails) (juju.CharmResources, error) {
+	jujuResources := make(juju.CharmResources, len(planResources))
+	for name, resource := range planResources {
+		var charmRevision, ociImageURL, registryUser, registryPassword string
+
+		if resource == "" {
+			return nil, fmt.Errorf("resource for %q is an empty string", name)
+		}
+
+		if _, err := strconv.Atoi(resource); err == nil {
+			charmRevision = resource
+		} else {
+			// Registry path matching is done based on a partial match.
+			// An image with URL "registry.example.com:5000/path/image:tag"
+			// will match a registry entry with key "registry.example.com:5000/path"
+			// but not "registry.example.com:5000" nor "registry.example.com".
+			//
+			// This allows for different credentials to be used for different paths
+			// within the same registry host e.g. github.com/canonical vs github.com/juju.
+			// This ASSUMES that multiple images from the same org/namespace can use the same credentials.
+			//
+			// If no match is found, no authentication is used.
+			ociImageURL = resource
+			imageIndex := strings.LastIndex(ociImageURL, "/")
+			registryPath := ociImageURL[:imageIndex]
+			if registryDetails, ok := imageRegistries[registryPath]; ok {
+				if !registryDetails.User.IsNull() {
+					registryUser = registryDetails.User.ValueString()
+				}
+				if !registryDetails.Password.IsNull() {
+					registryPassword = registryDetails.Password.ValueString()
+				}
+			}
+		}
+
+		jujuResources[name] = juju.CharmResource{
```

Comment:
On my point from earlier, if we don't just marshal the thing, you could have explicit types defined here, NewCharmResource could create the concrete type of a file resource, or oci image resource, and then shove them into the map here where the key is an interface and a .Type to figure what it is. It'd also stop you needing to do if cr.OCIImageURL